File: tests-ts/src/__tests__/formalClaimAlgebra.test.ts
/**
 * FACP-018: executable TypeScript Formal Claim Algebra binding tests.
 *
 * Acceptance:
 * - TypeScript accepts/rejects every normative vector.
 * - Browser-supplied policy/consent fields cannot construct authority.valid
 *   or effect.observed (browser-safe projection).
 * - Illegal transitions cannot construct a success type through public APIs.
 */

import { describe, it, expect } from 'vitest';
import {
  VOCAB_SCHEMA,
  RULES_SCHEMA,
  TASK_ID,
  GOAL_ID,
  BUNDLE,
  UNKNOWN_TRANSITION_POLICY,
  DIMENSION_ORDER,
  PREDICATE_ORDER,
  ORIGIN_VALUES,
  INTEGRITY_VALUES,
  AUTHORITY_VALUES,
  POLICY_VALUES,
  PROOF_VALUES,
  FRESHNESS_VALUES,
  EFFECT_VALUES,
  ENVIRONMENT_VALUES,
  REVIEW_VALUES,
  CLOSED_OUTCOME_VALUES,
  ORIGIN_ALLOWED,
  ORIGIN_FORBIDDEN,
  INTEGRITY_ALLOWED,
  AUTHORITY_ALLOWED,
  AUTHORITY_FORBIDDEN,
  POLICY_ALLOWED,
  POLICY_FORBIDDEN,
  PROOF_ALLOWED,
  PROOF_FORBIDDEN,
  FRESHNESS_ALLOWED,
  FRESHNESS_FORBIDDEN,
  EFFECT_ALLOWED,
  EFFECT_FORBIDDEN,
  ENVIRONMENT_ALLOWED,
  ENVIRONMENT_FORBIDDEN,
  REVIEW_ALLOWED,
  parseOrigin,
  parseIntegrity,
  parseAuthority,
  parsePolicy,
  parseProof,
  parseFreshness,
  parseEffect,
  parseEnvironment,
  parseReview,
  parseClosedOutcome,
  parsePromotionPredicate,
  weakestEnvelope,
  strongProductEnvelope,
  envelopeToCanonicalJson,
  envelopeFromCanonicalJson,
  envelopeFromDimensionMap,
  envelopeToDimensionMap,
  EvidenceBag,
  transitionAllowed,
  applyTransition,
  productionSupportedDimensions,
  effectSuccessfulDimensions,
  proofReusableDimensions,
  receiptAuthoritativeDimensions,
  releaseAdmissibleDimensions,
  predicateHolds,
  necessaryEvidence,
  ProductionSuccessClaim,
  VerifiedClaim,
  normativeVectors,
  evaluateNormativeVector,
  formatFcaError,
  projectBrowserSafeEnvelope,
  rejectBrowserAuthorityOrObservation,
  tryConstructFromBrowserFields,
  BROWSER_CLAIM_TOKENS,
} from '../formalClaimAlgebra.js';

describe('FACP-018 TypeScript FCA binding', () => {
  it('kernel identity matches FACP-018', () => {
    expect(VOCAB_SCHEMA).toBe('facp/formal-claim-algebra-v1@1');
    expect(RULES_SCHEMA).toBe('facp/promotion-rules@1');
    expect(TASK_ID).toBe('FACP-018');
    expect(GOAL_ID).toBe('FACP-G120');
    expect(BUNDLE).toBe('facp/fca/typescript');
    expect(UNKNOWN_TRANSITION_POLICY).toBe('reject');
    expect([...DIMENSION_ORDER]).toEqual([
      'origin',
      'integrity',
      'authority',
      'policy',
      'proof',
      'freshness',
      'effect',
      'environment',
      'review',
    ]);
    expect([...PREDICATE_ORDER]).toEqual([
      'production_supported',
      'effect_successful',
      'proof_reusable',
      'receipt_authoritative',
      'release_admissible',
    ]);
  });

  it('closed carriers reject unknown spellings', () => {
    expect(parseOrigin('live_observed').ok).toBe(true);
    expect(parseOrigin('LIVE_OBSERVED').ok).toBe(false);
    expect(parseIntegrity('maybe_valid').ok).toBe(false);
    expect(parseAuthority('payment').ok).toBe(false);
    expect(parsePolicy('browser_allow').ok).toBe(false);
    expect(parseProof('proven').ok).toBe(false);
    expect(parseFreshness('fresh').ok).toBe(false);
    expect(parseEffect('success').ok).toBe(false);
    expect(parseEnvironment('prod').ok).toBe(false);
    expect(parseReview('peer_reviewed').ok).toBe(false);
    expect(parseClosedOutcome('Success').ok).toBe(false);
    expect(parsePromotionPredicate('supported').ok).toBe(false);
  });

  it('carrier name tables are exhaustive', () => {
    expect(ORIGIN_VALUES).toHaveLength(6);
    expect(INTEGRITY_VALUES).toHaveLength(4);
    expect(AUTHORITY_VALUES).toHaveLength(6);
    expect(POLICY_VALUES).toHaveLength(5);
    expect(PROOF_VALUES).toHaveLength(6);
    expect(FRESHNESS_VALUES).toHaveLength(4);
    expect(EFFECT_VALUES).toHaveLength(7);
    expect(ENVIRONMENT_VALUES).toHaveLength(3);
    expect(REVIEW_VALUES).toHaveLength(3);
    expect(CLOSED_OUTCOME_VALUES).toHaveLength(9);
    for (const name of ORIGIN_VALUES) {
      const parsed = parseOrigin(name);
      expect(parsed.ok).toBe(true);
      if (parsed.ok) expect(parsed.value).toBe(name);
    }
    for (const name of PROOF_VALUES) {
      const parsed = parseProof(name);
      expect(parsed.ok).toBe(true);
      if (parsed.ok) expect(parsed.value).toBe(name);
    }
    for (const name of CLOSED_OUTCOME_VALUES) {
      const parsed = parseClosedOutcome(name);
      expect(parsed.ok).toBe(true);
      if (parsed.ok) expect(parsed.value).toBe(name);
    }
  });

  it('canonical serialization round-trips and rejects unknown fields', () => {
    const envelope = strongProductEnvelope();
    const json = envelopeToCanonicalJson(envelope);
    const parsed = envelopeFromCanonicalJson(json);
    expect(parsed.ok).toBe(true);
    if (parsed.ok) expect(parsed.value).toEqual(envelope);

    const bad =
      '{"origin":"absent","integrity":"unchecked","authority":"unchecked","policy":"unchecked","proof":"none","freshness":"stale","effect":"not_started","environment":"hermetic","review":"unreviewed","extra":true}';
    expect(envelopeFromCanonicalJson(bad).ok).toBe(false);

    const map = envelopeToDimensionMap(envelope);
    map['bogus'] = 'value';
    const unknown = envelopeFromDimensionMap(map);
    expect(unknown.ok).toBe(false);
    if (!unknown.ok) expect(unknown.error.kind).toBe('unknown_field');

    const incomplete: Record<string, string> = { origin: 'absent' };
    const missing = envelopeFromDimensionMap(incomplete);
    expect(missing.ok).toBe(false);
    if (!missing.ok) expect(missing.error.kind).toBe('missing_field');
  });

  it('every normative vector accepts or rejects as expected', () => {
    const vectors = normativeVectors();
    expect(vectors.length).toBeGreaterThanOrEqual(65);
    let accept = 0;
    let reject = 0;
    for (const vector of vectors) {
      const result = evaluateNormativeVector(vector);
      expect(result.ok, `${vector.id}: ${result.ok ? '' : formatFcaError(result.error)}`).toBe(
        true,
      );
      if (vector.expectation === 'accept') accept += 1;
      else reject += 1;
    }
    expect(accept).toBeGreaterThanOrEqual(40);
    expect(reject).toBeGreaterThanOrEqual(20);
  });

  it('illegal transitions cannot construct production success', () => {
    const bag = EvidenceBag.allNormative();

    const fixture = { ...strongProductEnvelope(), origin: 'fixture' as const };
    expect(transitionAllowed('origin', 'fixture', 'live_observed', bag).ok).toBe(false);
    expect(ProductionSuccessClaim.tryAdmit(fixture, bag).ok).toBe(false);
    expect(VerifiedClaim.tryAdmit(fixture, bag).ok).toBe(false);

    const hermetic = weakestEnvelope();
    expect(applyTransition(hermetic, 'environment', 'live', bag).ok).toBe(false);

    const expired = { ...strongProductEnvelope(), authority: 'expired' as const };
    expect(applyTransition(expired, 'authority', 'valid', bag).ok).toBe(false);
    expect(ProductionSuccessClaim.tryAdmit(expired, bag).ok).toBe(false);

    const stale = { ...strongProductEnvelope(), freshness: 'stale' as const };
    expect(applyTransition(stale, 'freshness', 'current', bag).ok).toBe(false);
    expect(ProductionSuccessClaim.tryAdmit(stale, bag).ok).toBe(false);

    const empty = new EvidenceBag();
    expect(transitionAllowed('proof', 'candidate', 'verified', empty).ok).toBe(false);
    const withVerifier = EvidenceBag.fromKeys([
      'named_current_verifier',
      'verifier_admission_closure',
    ]);
    expect(transitionAllowed('proof', 'candidate', 'verified', withVerifier).ok).toBe(true);

    const strong = strongProductEnvelope();
    const success = ProductionSuccessClaim.tryAdmit(strong, bag);
    expect(success.ok).toBe(true);
    if (success.ok) {
      expect(success.value.outcome()).toBe('Verified');
      expect(success.value.envelope()).toEqual(strong);
    }
    const verified = VerifiedClaim.tryAdmit(strong, bag);
    expect(verified.ok).toBe(true);
    if (verified.ok) {
      expect(verified.value.outcome()).toBe('Verified');
      expect(verified.value.envelope()).toEqual(strong);
    }
  });

  it('dimension predicate parity with strong and weak envelopes', () => {
    const strong = strongProductEnvelope();
    expect(productionSupportedDimensions(strong)).toBe(true);
    expect(effectSuccessfulDimensions(strong)).toBe(true);
    expect(proofReusableDimensions(strong)).toBe(true);
    expect(receiptAuthoritativeDimensions(strong)).toBe(true);
    expect(releaseAdmissibleDimensions(strong)).toBe(true);

    const weak = weakestEnvelope();
    expect(productionSupportedDimensions(weak)).toBe(false);
    expect(effectSuccessfulDimensions(weak)).toBe(false);
    expect(proofReusableDimensions(weak)).toBe(false);
    expect(receiptAuthoritativeDimensions(weak)).toBe(false);
    expect(releaseAdmissibleDimensions(weak)).toBe(false);

    const bag = EvidenceBag.allNormative();
    for (const pred of PREDICATE_ORDER) {
      expect(predicateHolds(pred, strong, bag).ok).toBe(true);
      expect(predicateHolds(pred, weak, bag).ok).toBe(false);
      expect(necessaryEvidence(pred).length).toBeGreaterThan(0);
    }
  });

  it('transition table counts match promotion-rules parity', () => {
    expect(ORIGIN_ALLOWED).toHaveLength(8);
    expect(ORIGIN_FORBIDDEN).toHaveLength(4);
    expect(INTEGRITY_ALLOWED).toHaveLength(6);
    expect(AUTHORITY_ALLOWED).toHaveLength(9);
    expect(AUTHORITY_FORBIDDEN).toHaveLength(3);
    expect(POLICY_ALLOWED).toHaveLength(8);
    expect(POLICY_FORBIDDEN).toHaveLength(1);
    expect(PROOF_ALLOWED).toHaveLength(12);
    expect(PROOF_FORBIDDEN).toHaveLength(1);
    expect(FRESHNESS_ALLOWED).toHaveLength(6);
    expect(FRESHNESS_FORBIDDEN).toHaveLength(4);
    expect(EFFECT_ALLOWED).toHaveLength(11);
    expect(EFFECT_FORBIDDEN).toHaveLength(2);
    expect(ENVIRONMENT_ALLOWED).toHaveLength(2);
    expect(ENVIRONMENT_FORBIDDEN).toHaveLength(2);
    expect(REVIEW_ALLOWED).toHaveLength(3);
  });

  it('browser-safe projection cannot construct authority or observation', () => {
    const fields = {
      policy: { outcome: 'allow' },
      consent: 'granted',
      allow: true,
      dry_run: false,
      policy_decision: { outcome: 'allow' },
      confirmation_token: 'ui-token',
    };

    const projected = projectBrowserSafeEnvelope(fields);
    expect(projected.authority).not.toBe('valid');
    expect(projected.policy).not.toBe('allowed');
    expect(projected.policy).not.toBe('allowed_with_obligations');
    expect(projected.effect).not.toBe('observed');
    expect(projected.authority).toBe('unchecked');
    expect(projected.policy).toBe('unchecked');
    expect(projected.effect).toBe('not_started');

    // Differing only by browser allow/consent still yields identical admission dims.
    const denied = projectBrowserSafeEnvelope({ consent: 'denied', allow: false });
    expect(denied.authority).toBe(projected.authority);
    expect(denied.policy).toBe(projected.policy);
    expect(denied.effect).toBe(projected.effect);

    const rejectAuth = rejectBrowserAuthorityOrObservation({
      fields,
      requestedAuthority: 'valid',
    });
    expect(rejectAuth.ok).toBe(false);
    if (!rejectAuth.ok) {
      expect(rejectAuth.error.kind).toBe('browser_projection_rejected');
      expect(formatFcaError(rejectAuth.error)).toMatch(/authority\.valid|NONIMP_/);
    }

    const rejectObs = rejectBrowserAuthorityOrObservation({
      fields,
      requestedEffect: 'observed',
    });
    expect(rejectObs.ok).toBe(false);
    if (!rejectObs.ok) {
      expect(formatFcaError(rejectObs.error)).toMatch(/effect\.observed|NONIMP_/);
    }

    const rejectPol = rejectBrowserAuthorityOrObservation({
      fields,
      claimTokens: ['browser_policy', 'browser_consent'],
      requestedPolicy: 'allowed',
    });
    expect(rejectPol.ok).toBe(false);
    if (!rejectPol.ok) {
      expect(rejectPol.error.code).toBe('NONIMP_BROWSER_TO_HOST_POLICY');
    }

    const constructAuth = tryConstructFromBrowserFields(fields, {
      authority: 'valid',
      policy: 'allowed',
      effect: 'observed',
    });
    expect(constructAuth.ok).toBe(false);

    const safeOnly = tryConstructFromBrowserFields(fields, {});
    expect(safeOnly.ok).toBe(true);
    if (safeOnly.ok) {
      expect(safeOnly.value.authority).toBe('unchecked');
      expect(safeOnly.value.effect).toBe('not_started');
      expect(safeOnly.value.policy).toBe('unchecked');
    }

    // Browser claim tokens alone cannot promote host policy.
    for (const token of BROWSER_CLAIM_TOKENS) {
      const r = rejectBrowserAuthorityOrObservation({
        claimTokens: [token],
        requestedPolicy: 'allowed',
      });
      expect(r.ok).toBe(false);
    }

    // Success types remain unreachable from browser projection + empty evidence.
    expect(ProductionSuccessClaim.tryAdmit(projected, new EvidenceBag()).ok).toBe(false);
    expect(VerifiedClaim.tryAdmit(projected, EvidenceBag.allNormative()).ok).toBe(false);
  });
});

File: tests-rs/tests/formal_claim_algebra_test.rs
//! FACP-013: executable Rust Formal Claim Algebra kernel tests.
//!
//! Acceptance:
//! - Rust accepts/rejects every normative vector.
//! - Illegal transitions cannot construct a success type through public APIs.
//! - Exhaustive closed carriers, transition parity, and canonical serialization.
//!
//! `lib.rs` is outside this task's edit scope, so the kernel module is loaded
//! via path include from `src/formal_claim_algebra.rs`.

#![forbid(unsafe_code)]

#[path = "../src/formal_claim_algebra.rs"]
mod formal_claim_algebra;

use formal_claim_algebra::*;
use std::collections::BTreeMap;
use std::str::FromStr;

#[test]
fn kernel_identity_matches_facp_013() {
    assert_eq!(VOCAB_SCHEMA, "facp/formal-claim-algebra-v1@1");
    assert_eq!(RULES_SCHEMA, "facp/promotion-rules@1");
    assert_eq!(TASK_ID, "FACP-013");
    assert_eq!(GOAL_ID, "FACP-G120");
    assert_eq!(BUNDLE, "facp/fca/rust");
    assert_eq!(UNKNOWN_TRANSITION_POLICY, "reject");
    assert_eq!(
        DIMENSION_ORDER,
        [
            "origin",
            "integrity",
            "authority",
            "policy",
            "proof",
            "freshness",
            "effect",
            "environment",
            "review"
        ]
    );
    assert_eq!(
        PREDICATE_ORDER,
        [
            "production_supported",
            "effect_successful",
            "proof_reusable",
            "receipt_authoritative",
            "release_admissible"
        ]
    );
}

#[test]
fn closed_carriers_reject_unknown_spellings() {
    assert!(Origin::from_str("live_observed").is_ok());
    assert!(matches!(
        Origin::from_str("LIVE_OBSERVED"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Integrity::from_str("maybe_valid"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Authority::from_str("payment"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Policy::from_str("browser_allow"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Proof::from_str("proven"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Freshness::from_str("fresh"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Effect::from_str("success"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Environment::from_str("prod"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        Review::from_str("peer_reviewed"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        ClosedOutcome::from_str("Success"),
        Err(FcaError::UnknownEnum { .. })
    ));
    assert!(matches!(
        PromotionPredicate::from_str("supported"),
        Err(FcaError::UnknownEnum { .. })
    ));
}

#[test]
fn carrier_name_tables_are_exhaustive() {
    assert_eq!(Origin::names().len(), 6);
    assert_eq!(Integrity::names().len(), 4);
    assert_eq!(Authority::names().len(), 6);
    assert_eq!(Policy::names().len(), 5);
    assert_eq!(Proof::names().len(), 6);
    assert_eq!(Freshness::names().len(), 4);
    assert_eq!(Effect::names().len(), 7);
    assert_eq!(Environment::names().len(), 3);
    assert_eq!(Review::names().len(), 3);
    assert_eq!(ClosedOutcome::names().len(), 9);
    for name in Origin::names() {
        assert_eq!(Origin::from_str(name).unwrap().as_str(), *name);
    }
    for name in Proof::names() {
        assert_eq!(Proof::from_str(name).unwrap().as_str(), *name);
    }
    for name in ClosedOutcome::names() {
        assert_eq!(ClosedOutcome::from_str(name).unwrap().as_str(), *name);
    }
}

#[test]
fn canonical_serialization_round_trip_and_unknown_field_rejection() {
    let envelope = EvidenceEnvelope::strong_product();
    let json = envelope.to_canonical_json().expect("serialize");
    let parsed = EvidenceEnvelope::from_canonical_json(&json).expect("parse");
    assert_eq!(parsed, envelope);

    let bad = r#"{"origin":"absent","integrity":"unchecked","authority":"unchecked","policy":"unchecked","proof":"none","freshness":"stale","effect":"not_started","environment":"hermetic","review":"unreviewed","extra":true}"#;
    assert!(EvidenceEnvelope::from_canonical_json(bad).is_err());

    let mut map = envelope.to_dimension_map();
    map.insert("bogus".into(), "value".into());
    assert!(matches!(
        EvidenceEnvelope::from_dimension_map(&map),
        Err(FcaError::UnknownField(_))
    ));

    let mut incomplete: BTreeMap<String, String> = BTreeMap::new();
    incomplete.insert("origin".into(), "absent".into());
    assert!(matches!(
        EvidenceEnvelope::from_dimension_map(&incomplete),
        Err(FcaError::MissingField(_))
    ));
}

#[test]
fn every_normative_vector_accepts_or_rejects_as_expected() {
    let vectors = normative_vectors();
    assert!(
        vectors.len() >= 65,
        "expected full accept+reject suite, got {}",
        vectors.len()
    );
    let mut accept = 0usize;
    let mut reject = 0usize;
    for vector in &vectors {
        evaluate_normative_vector(vector).unwrap_or_else(|e| panic!("{e}"));
        match vector.expectation {
            VectorExpectation::Accept => accept += 1,
            VectorExpectation::Reject => reject += 1,
        }
    }
    assert!(accept >= 40, "accept count too low: {accept}");
    assert!(reject >= 20, "reject count too low: {reject}");
}

#[test]
fn illegal_transitions_cannot_construct_production_success() {
    let bag = EvidenceBag::all_normative();

    // Fixture cannot transition to live_observed, and cannot admit success.
    let fixture = EvidenceEnvelope {
        origin: Origin::Fixture,
        ..EvidenceEnvelope::strong_product()
    };
    assert!(transition_allowed(
        Dimension::Origin,
        "fixture",
        "live_observed",
        &bag
    )
    .is_err());
    assert!(ProductionSuccessClaim::try_admit(fixture.clone(), &bag).is_err());
    assert!(VerifiedClaim::try_admit(fixture, &bag).is_err());

    // Hermetic → live is absolutely forbidden even with live evidence.
    let hermetic = EvidenceEnvelope::weakest();
    assert!(apply_transition(&hermetic, Dimension::Environment, "live", &bag).is_err());

    // Expired authority cannot relabel to valid; success type stays unreachable.
    let expired = EvidenceEnvelope {
        authority: Authority::Expired,
        ..EvidenceEnvelope::strong_product()
    };
    assert!(apply_transition(&expired, Dimension::Authority, "valid", &bag).is_err());
    assert!(ProductionSuccessClaim::try_admit(expired, &bag).is_err());

    // Stale freshness cannot become current.
    let stale = EvidenceEnvelope {
        freshness: Freshness::Stale,
        ..EvidenceEnvelope::strong_product()
    };
    assert!(apply_transition(&stale, Dimension::Freshness, "current", &bag).is_err());
    assert!(ProductionSuccessClaim::try_admit(stale, &bag).is_err());

    // Candidate proof cannot become verified without verifier evidence.
    let empty = EvidenceBag::new();
    assert!(transition_allowed(Dimension::Proof, "candidate", "verified", &empty).is_err());
    let with_verifier = EvidenceBag::from_keys([
        "named_current_verifier",
        "verifier_admission_closure",
    ]);
    assert!(transition_allowed(
        Dimension::Proof,
        "candidate",
        "verified",
        &with_verifier
    )
    .is_ok());

    // Strong product + full evidence admits gated success types.
    let strong = EvidenceEnvelope::strong_product();
    let success = ProductionSuccessClaim::try_admit(strong.clone(), &bag).expect("admit");
    assert_eq!(success.outcome(), ClosedOutcome::Verified);
    assert_eq!(success.envelope(), &strong);
    let verified = VerifiedClaim::try_admit(strong.clone(), &bag).expect("verified");
    assert_eq!(verified.outcome(), ClosedOutcome::Verified);
    assert_eq!(verified.envelope(), &strong);

    let mut bag2 = EvidenceBag::new();
    bag2.insert("named_current_verifier");
    bag2.insert("verifier_admission_closure");
    assert!(bag2.contains_all(&["named_current_verifier", "verifier_admission_closure"]));
}

#[test]
fn dimension_predicate_parity_with_strong_and_weak_envelopes() {
    let strong = EvidenceEnvelope::strong_product();
    assert!(production_supported_dimensions(&strong));
    assert!(effect_successful_dimensions(&strong));
    assert!(proof_reusable_dimensions(&strong));
    assert!(receipt_authoritative_dimensions(&strong));
    assert!(release_admissible_dimensions(&strong));

    let weak = EvidenceEnvelope::weakest();
    assert!(!production_supported_dimensions(&weak));
    assert!(!effect_successful_dimensions(&weak));
    assert!(!proof_reusable_dimensions(&weak));
    assert!(!receipt_authoritative_dimensions(&weak));
    assert!(!release_admissible_dimensions(&weak));

    for pred in PromotionPredicate::all() {
        assert!(predicate_holds(pred, &strong, &EvidenceBag::all_normative()).is_ok());
        assert!(predicate_holds(pred, &weak, &EvidenceBag::all_normative()).is_err());
    }
}

#[test]
fn exhaustive_match_smoke_for_closed_outcomes_and_predicates() {
    // Exhaustive matches: adding a variant without updating these arms fails to compile.
    for outcome in [
        ClosedOutcome::Unavailable,
        ClosedOutcome::Rejected,
        ClosedOutcome::Simulated,
        ClosedOutcome::Attempted,
        ClosedOutcome::Unknown,
        ClosedOutcome::Observed,
        ClosedOutcome::Verified,
        ClosedOutcome::Failed,
        ClosedOutcome::Compensated,
    ] {
        let _ = match outcome {
            ClosedOutcome::Unavailable
            | ClosedOutcome::Rejected
            | ClosedOutcome::Simulated
            | ClosedOutcome::Attempted
            | ClosedOutcome::Unknown
            | ClosedOutcome::Observed
            | ClosedOutcome::Verified
            | ClosedOutcome::Failed
            | ClosedOutcome::Compensated => outcome.as_str(),
        };
    }
    for pred in PromotionPredicate::all() {
        let _ = match pred {
            PromotionPredicate::ProductionSupported
            | PromotionPredicate::EffectSuccessful
            | PromotionPredicate::ProofReusable
            | PromotionPredicate::ReceiptAuthoritative
            | PromotionPredicate::ReleaseAdmissible => necessary_evidence(pred),
        };
    }
}

#[test]
fn transition_table_counts_match_promotion_rules_parity() {
    assert_eq!(ORIGIN_ALLOWED.len(), 8);
    assert_eq!(ORIGIN_FORBIDDEN.len(), 4);
    assert_eq!(INTEGRITY_ALLOWED.len(), 6);
    assert_eq!(AUTHORITY_ALLOWED.len(), 9);
    assert_eq!(AUTHORITY_FORBIDDEN.len(), 3);
    assert_eq!(POLICY_ALLOWED.len(), 8);
    assert_eq!(POLICY_FORBIDDEN.len(), 1);
    assert_eq!(PROOF_ALLOWED.len(), 12);
    assert_eq!(PROOF_FORBIDDEN.len(), 1);
    assert_eq!(FRESHNESS_ALLOWED.len(), 6);
    assert_eq!(FRESHNESS_FORBIDDEN.len(), 4);
    assert_eq!(EFFECT_ALLOWED.len(), 11);
    assert_eq!(EFFECT_FORBIDDEN.len(), 2);
    assert_eq!(ENVIRONMENT_ALLOWED.len(), 2);
    assert_eq!(ENVIRONMENT_FORBIDDEN.len(), 2);
    assert_eq!(REVIEW_ALLOWED.len(), 3);
}
